### DISCUSSION TITLE
Guides to make minor visual changes on Budibase code?

### DISCUSSION THREAD

**Discussion by magickaito** [Budibase/budibase#6218] [Help] [answered]
I had a lot of fun trying out the builder on self-hosted digitalocean. 

Now I am trying to make minor changes to svelte files, rebuild it on digitalocean. 

For example, thinking of removing this question mark at the bottom right of the app selection page:

![Screenshot 2022-06-07 at 6 15 34 PM](https://user-images.githubusercontent.com/1195573/172356514-b721799f-8051-4c07-bb28-9e23773ee4fb.jpg)

It appears that I need to edit App.svelte and recompile the whole thing.

![Screenshot 2022-06-07 at 5 54 02 PM](https://user-images.githubusercontent.com/1195573/172356242-ae83d6f5-14a4-4d07-b26e-366c2cff3fd8.jpg)

Any step by step guides on how to achieve this?

**Comment by melohagan** [comment]
Hi @magickaito 

Looks like you've found it - you could just delete the help icon component.

This change wouldn't be accepted into the main branch, so it'd only be useful for your own fork.

**Reply by magickaito** [reply]
@melohagan Actually, I have no idea where is the file on the 1-clicked digitalocean image at this point of time, and how to rebuild the whole thing on digitalocean if I make the changes? Is there a beginner guide on this xD. The closest guide I can find so far is the https://github.com/Budibase/budibase/blob/HEAD/.github/CONTRIBUTING.md which is not really about editing and rebuilding the 1-clicked image either. Hmm....

**Reply by VikingDadMedic** [reply]
@magickaito I have customized a couple of aspects of BB. That being said, no surprise, deleting/removing things is a synch… while adding/modifying (e.g., like adding options to the rich text editor - EasyMDE) can become more complex depending on your requirements.

Either way - whether you using 1 click images or not - the pipeline is almost identical (until you get to the actual deployment provider). And while there is always more than one path, the fundamental steps are going to be:

1.) Fork the repo (make sure it’s not the development branch but instead the latest release)

2.) [Optional / Recommended] Open the package.json et all config files. Inspect all the scripts to get an idea of how the code base is set up, 5 minutes looking at all the commands can save you major time later when it comes to rebuilding. 

3.) Make all the changes you want (though I’d recommend just doing a quick POC test and then jump to the next step). 

4.) Now what you’re going to want to do is look at the [Docker files](https://github.com/Budibase/budibase/tree/b0b69222d1435451dffc068b17c0776c044c88ab/packages/worker) and [hosting files](https://github.com/Budibase/budibase/tree/3275a9efb8b082872ecb8f865bbbda13d23261e4/hosting). 

The general concept being, if you make changes to any of the sources files, your going to have to swap out that part of the pipeline in the corresponding docker/hosting files for the changes to take place. 

5.) Once you have your new docker image(s) and other files - you need to push those to a container registry. Digital ocean makes this somewhat easy ([quick start](https://docs.digitalocean.com/products/container-registry/quickstart/)). 

6.) When you make it that far, you’re pretty much home free - just follow the D.O. steps for deployment like you did before using your new replacement images where needed.

I know this is a high level over simplified explanation but it can get complex if I tried to get into details (especially without knowing how extensive or what your customization will be). 

Best of luck!

**Reply by magickaito** [reply]
Thanks so much for the time to explain to a Docker noob like me. So if I want to change one line of code in packages/builder/src/App.svelte, the general concept is that I need to swap out the app-service in docker-compose.yaml to a forked image? something like that? xD

**Reply by SantoshSrinivas79** [reply]
Thanks mate. Docker noob here too! .... Would love to get exact commands to run after making the code changes :-) (like after opening terminal and `cd` :) ... thats the level of noobness )

**Reply by magickaito** [reply]
> Thanks mate. Docker noob here too! .... Would love to get exact commands to run after making the code changes :-) (like after opening terminal and `cd` :) ... thats the level of noobness )

haha me too+1

**Comment by magickaito** [comment, marked as answer]
Finally successfully deployed my slightly modified Budibase, following this Single Docker Image instruction:
https://github.com/Budibase/budibase/tree/develop/hosting/single

Generally:
- First fork the repo, checkout master branch
- Then create a branch at a recent master branch commit (I succeeded at version 1.1.7)
- Make your necessary changes in codebase. Commit and push this branch to your forked repo.
- Create a DigitalOcean droplet. SSH into your droplet
- Follow all the steps in Single Docker Image instruction above (Except, clone from your forked repo)
- Your changes will now up and running on the droplet

Additional notes on cloning your forked repo on the fresh DO droplet. 
- Generate SSH key pairs on the droplet
`ssh-keygen -t ed25519 -C "your@email.com"`
- cd into /root/.ssh, search for the id_ed25519.pub to copy the content generated
- goto [github.com](http://github.com/) settings → SSH and GPG Keys, click new SSH key. Paste the public key copied in previous step. 
- now you can clone your forked repo:
`git clone --branch yourbranchwithchanges git@github.com:yourusername/budibase.git`

**Reply by magickaito** [reply]
@SantoshSrinivas79

**Reply by shogunpurple** [reply]
Hi @magickaito - thank you for sharing, this will be super helpful for the community. Out of interest, just wondering what sort of modifications you were making to budibase (apart from the help icon), and if it's something that we could possibly improve in the product itself.

**Reply by SantoshSrinivas79** [reply]
Thank you @magickaito! I will try this first thing morning! :) .... would love to try out primarily styling changes

**Reply by magickaito** [reply]
@SantoshSrinivas79 u r welcome! 
@shogunpurple thanks for the fantastic product. As a start I may wish to: 1) replace the bottom right help link to some internal feedback email 2) change the website title and favicon to match client's organisation https://github.com/Budibase/budibase/discussions/2897 3) hide some options for non-technical admins, e.g. let them manage users but don't let them edit apps. https://github.com/Budibase/budibase/discussions/6417

**Comment by poirazis** [comment]
This is very helpful @magickaito ! Thank you! 
We should start building some more Components now!
